File: apps/admin-web/lib/features/organisations/data/admin_organisations_repository.dart
import 'package:dio/dio.dart';
import 'package:vitacare_shared/vitacare_shared.dart';
import '../../../core/network/api_exception.dart';

class AdminOrganisationListItem {
  final String userId;
  final int? orgNumber;
  final String fullName;
  final String phone;
  final String organisationName;
  final String organisationType;
  final String city;
  final String area;
  final bool isActive;
  final bool isJobPostingBlocked;
  final String? blockReason;
  final String createdAt;

  const AdminOrganisationListItem({
    required this.userId,
    this.orgNumber,
    required this.fullName,
    required this.phone,
    required this.organisationName,
    required this.organisationType,
    required this.city,
    required this.area,
    required this.isActive,
    required this.isJobPostingBlocked,
    this.blockReason,
    required this.createdAt,
  });

  factory AdminOrganisationListItem.fromJson(Map<String, dynamic> json) => AdminOrganisationListItem(
        userId: json['user_id'] as String,
        orgNumber: json['org_number'] as int?,
        fullName: json['full_name'] as String,
        phone: json['phone'] as String,
        organisationName: json['organisation_name'] as String,
        organisationType: json['organisation_type'] as String,
        city: json['city'] as String,
        area: json['area'] as String,
        isActive: json['is_active'] as bool,
        isJobPostingBlocked: json['is_job_posting_blocked'] as bool,
        blockReason: json['block_reason'] as String?,
        createdAt: json['created_at'] as String,
      );
}

class AdminOrganisationsListResult {
  final List<AdminOrganisationListItem> items;
  final PaginationMeta meta;

  const AdminOrganisationsListResult({required this.items, required this.meta});
}

/// Mirrors AdminIndividualsRepository exactly.
class AdminOrganisationsRepository {
  final Dio _dio;

  AdminOrganisationsRepository(this._dio);

  Future<AdminOrganisationsListResult> list({int page = 1, int limit = 20}) async {
    try {
      final res = await _dio.get('/admin/organisations', queryParameters: {'page': page, 'limit': limit});
      final items = (res.data['data'] as List)
          .map((json) => AdminOrganisationListItem.fromJson(json as Map<String, dynamic>))
          .toList();
      final meta = PaginationMeta.fromJson(res.data['meta'] as Map<String, dynamic>);
      return AdminOrganisationsListResult(items: items, meta: meta);
    } on DioException catch (e) {
      throw ApiException.fromDioException(e);
    }
  }

  /// [level] is 'job_posting' (blocks only new postings) or 'full' (login
  /// lockout, reuses users.is_active).
  Future<void> block(String userId, String level, String reason) async {
    try {
      await _dio.patch('/admin/organisations/$userId/block', data: {'level': level, 'reason': reason});
    } on DioException catch (e) {
      throw ApiException.fromDioException(e);
    }
  }

  Future<void> unblock(String userId, String level) async {
    try {
      await _dio.patch('/admin/organisations/$userId/unblock', data: {'level': level});
    } on DioException catch (e) {
      throw ApiException.fromDioException(e);
    }
  }
}

File: apps/api/src/database/repositories/admin-organisations.repository.ts
import { Injectable } from '@nestjs/common';
import { DatabaseService } from '../database.service';
import { ListPage } from './jobs.repository';

export interface AdminOrganisationListItem {
  user_id: string;
  org_number: number;
  full_name: string;
  phone: string;
  organisation_name: string;
  organisation_type: string;
  city: string;
  area: string;
  is_active: boolean;
  is_job_posting_blocked: boolean;
  block_reason: string | null;
  created_at: Date;
}

export interface AdminOrganisationListFilters {
  /** Matches organisation_name, contact full_name, phone, or the display id
   *  (ORG-<n>). */
  search?: string;
  blockStatus?: 'active' | 'job_posting_blocked' | 'blocked';
  organisationType?: string;
  city?: string;
}

function buildOrganisationsWhereClause(filters: AdminOrganisationListFilters): {
  clause: string;
  params: unknown[];
} {
  const conditions = [`u.role = 'organisation'`];
  const params: unknown[] = [];
  if (filters.search) {
    params.push(`%${filters.search}%`);
    conditions.push(
      `(op.organisation_name ILIKE $${params.length} OR u.full_name ILIKE $${params.length}
        OR u.phone ILIKE $${params.length} OR ('ORG-' || op.org_number::text) ILIKE $${params.length})`,
    );
  }
  if (filters.blockStatus === 'blocked') {
    conditions.push(`u.is_active = false`);
  } else if (filters.blockStatus === 'job_posting_blocked') {
    conditions.push(`u.is_active = true AND op.is_job_posting_blocked = true`);
  } else if (filters.blockStatus === 'active') {
    conditions.push(`u.is_active = true AND op.is_job_posting_blocked = false`);
  }
  if (filters.organisationType) {
    params.push(filters.organisationType);
    conditions.push(`op.organisation_type = $${params.length}`);
  }
  if (filters.city) {
    params.push(filters.city);
    conditions.push(`op.city = $${params.length}`);
  }
  return { clause: `WHERE ${conditions.join(' AND ')}`, params };
}

/** Admin-side view of organisation (hospital/rehab/clinic) accounts —
 *  mirrors AdminIndividualsRepository exactly. Self-service lives in
 *  OrganisationProfilesRepository; admin listing/block actions live here. */
@Injectable()
export class AdminOrganisationsRepository {
  constructor(private readonly db: DatabaseService) {}

  async listOrganisations(
    filters: AdminOrganisationListFilters,
    page: ListPage,
  ): Promise<{ items: AdminOrganisationListItem[]; total: number }> {
    const { clause, params } = buildOrganisationsWhereClause(filters);
    const offset = (page.page - 1) * page.limit;
    const listParams = [...params, page.limit, offset];
    const limitPlaceholder = `$${listParams.length - 1}`;
    const offsetPlaceholder = `$${listParams.length}`;

    const [listResult, countResult] = await Promise.all([
      this.db.query<AdminOrganisationListItem>(
        `SELECT u.id AS user_id, op.org_number, u.full_name, u.phone,
                op.organisation_name, op.organisation_type, op.city, op.area,
                u.is_active, op.is_job_posting_blocked, op.block_reason, u.created_at
         FROM users u
         JOIN organisation_profiles op ON op.user_id = u.id
         ${clause}
         ORDER BY u.created_at DESC
         LIMIT ${limitPlaceholder} OFFSET ${offsetPlaceholder}`,
        listParams,
      ),
      this.db.query<{ count: string }>(
        `SELECT COUNT(*) FROM users u
         JOIN organisation_profiles op ON op.user_id = u.id
         ${clause}`,
        params,
      ),
    ]);
    return { items: listResult.rows, total: Number(countResult.rows[0].count) };
  }

  async findDetailByUserId(userId: string): Promise<AdminOrganisationListItem | null> {
    const result = await this.db.query<AdminOrganisationListItem>(
      `SELECT u.id AS user_id, op.org_number, u.full_name, u.phone,
              op.organisation_name, op.organisation_type, op.city, op.area,
              u.is_active, op.is_job_posting_blocked, op.block_reason, u.created_at
       FROM users u
       JOIN organisation_profiles op ON op.user_id = u.id
       WHERE u.id = $1 AND u.role = 'organisation'`,
      [userId],
    );
    return result.rows[0] ?? null;
  }

  async setJobPostingBlocked(userId: string, blocked: boolean, reason: string | null): Promise<void> {
    await this.db.query(
      `UPDATE organisation_profiles
       SET is_job_posting_blocked = $2, block_reason = $3, updated_at = NOW()
       WHERE user_id = $1`,
      [userId, blocked, reason],
    );
  }

  async setBlockReason(userId: string, reason: string | null): Promise<void> {
    await this.db.query(
      `UPDATE organisation_profiles SET block_reason = $2, updated_at = NOW() WHERE user_id = $1`,
      [userId, reason],
    );
  }

  /** Admin override — only fields present (not undefined) on `input` are
   *  written. Keyed on user_id (unlike caregiver_profiles' adminUpdate,
   *  which is keyed on the profile's own id) since that's what every
   *  other method on this repository already keys on. */
  async adminUpdate(
    userId: string,
    input: {
      organisation_name?: string;
      contact_person_name?: string;
      organisation_type?: string;
      city?: string;
      area?: string;
    },
  ): Promise<void> {
    const entries = Object.entries(input).filter(([, value]) => value !== undefined);
    if (entries.length === 0) return;

    const setClauses = entries.map(([key], i) => `${key} = $${i + 2}`);
    const values = entries.map(([, value]) => value);
    await this.db.query(
      `UPDATE organisation_profiles SET ${setClauses.join(', ')}, updated_at = NOW() WHERE user_id = $1`,
      [userId, ...values],
    );
  }
}
